library/general/src/lib/yast2/rel_url.rb

```diff
@@ -0,0 +1,130 @@
+# find current contact information at www.suse.com.
+
+require "yast"
+require "uri"
+
+Yast.import "InstURL"
+
+module Yast2
+  # Class for working with relative URLs ("relurl://")
+  class RelURL
+    attr_reader :base, :relative
+
+    # Is the URL a relative URL?
+    #
+    # @param url [String, URI] the URL
+    # @return [Boolean] `true` if the URL uses the "relurl" schema, otherwise `false`
+    def self.relurl?(url)
+      URI(url).scheme == "relurl"
+    end
+
+    # Create RelURL object with URL relative to the installation repository
+    #
+    # @param rel_url [String, URI] the relative URL
+    # @return [RelURL]
+    #
+    # @note Works properly only during installation/upgrade, do not use
+    #   in an installed system.
+    def self.from_installation_repository(rel_url)
+      base_url = Yast::InstURL.installInf2Url("")
+      new(base_url, rel_url)
+    end
+
+    # Constructor
+    #
+    # @param base_url [String,URI] the base URL
+    # @param rel_url [String,URI] the relative URL, it should use the "relurl://"
+    #  schema otherwise the base URL is ignored
+    def initialize(base_url, rel_url)
+      @base = import_url(base_url)
+      @relative = import_url(rel_url)
+    end
+
+    # Build and absolute URL
+    #
+    # @param path [String,nil] optional URL subpath
+    # @return [URI] the absolute URL
+    #
+    # @note It internally uses the Ruby `File.expand_path` function which
+    # also evaluates the parent directory path ("../") so it is possible
+    # to go up in the tree using the "relurl://../foo" or the "../foo" path
+    # parameter.
+    def absolute_url(path = nil)
+      if (!relative.to_s.empty? && !RelURL.relurl?(relative)) || base.to_s.empty?
+        ret = relative.dup
+        relative_url = URI("")
+      else
+        ret = base.dup
+        relative_url = relative.dup
+      end
+
+      relative_path = relative_url.path
+      relative_path = File.join(relative_path, path) if path && !path.empty?
+
+      base_path = ret.path
+      if !base_path.empty? || !relative_path.empty?
+        # the path would be expanded from the current working directory
+        # by File.expand_path if the base path is not absolute
+        base_path.prepend("/") if !base_path.start_with?("/")
+
+        # escape the "~"" character, it is treated as a home directory name by File.expand_path,
+        # moreover it raises ArgumentError if that user does not exist in the system
+        relative_path.gsub!("~", "%7E")
+        # the relative path really needs to be relative, remove the leading slash(es)
+        relative_path.sub!(/\A\/+/, "")
+
+        ret.path = File.expand_path(relative_path, base_path)
+      end
+
+      export_url(ret)
+      ret
+    end
+
+  private
+
+    # a helper method for importing an URL,
+    # it creates a copy of the input object and pre-processes the
+    # "file://" and "relurl://" URLs
+    def import_url(url)
```

Comment:
name is kind of strange, as it for me do more like read or parse and not import.